Src/FluentAssertions/Equivalency/ConstraintEquivalencyStep.cs

```diff
@@ -0,0 +1,228 @@
+                .FailWith("Expected {context:constraint} to be named {0}{reason}, but found {1}", expectation.ConstraintName, subject.ConstraintName);
+
+            parent.AssertEqualityUsing(
+                CreateNestedContext(context, nameof(subject.ExtendedProperties)));
+
+            if (selectedMembers.ContainsKey(nameof(expectation.RelatedTable)))
+            {
+                AssertionScope.Current
+                    .ForCondition(subject.RelatedTable.TableName == expectation.RelatedTable.TableName)
+                    .FailWith("Expected {context:constraint} to have a related table named {0}{reason}, but found {1}", expectation.RelatedTable.TableName, subject.RelatedTable.TableName);
+            }
+
+            if (selectedMembers.ContainsKey(nameof(expectation.AcceptRejectRule)))
+            {
+                AssertionScope.Current
+                    .ForCondition(subject.AcceptRejectRule == expectation.AcceptRejectRule)
+                    .FailWith("Expected {context:constraint} to have AcceptRejectRule.{0}{reason}, but found AcceptRejectRule.{1}", expectation.AcceptRejectRule, subject.AcceptRejectRule);
+            }
+
+            if (selectedMembers.ContainsKey(nameof(expectation.DeleteRule)))
+            {
+                AssertionScope.Current
+                    .ForCondition(subject.DeleteRule == expectation.DeleteRule)
+                    .FailWith("Expected {context:constraint} to have DeleteRule Rule.{0}{reason}, but found Rule.{1}", expectation.DeleteRule, subject.DeleteRule);
+            }
+
+            if (selectedMembers.ContainsKey(nameof(expectation.UpdateRule)))
+            {
+                AssertionScope.Current
+                    .ForCondition(subject.UpdateRule == expectation.UpdateRule)
+                    .FailWith("Expected {context:constraint} to have UpdateRule Rule.{0}{reason}, but found Rule.{1}", expectation.UpdateRule, subject.UpdateRule);
+            }
+
+            if (selectedMembers.ContainsKey(nameof(expectation.Columns)))
+            {
+                CompareConstraintColumns(subject.Columns, expectation.Columns);
+            }
+
+            if (selectedMembers.ContainsKey(nameof(expectation.RelatedColumns)))
+            {
+                CompareConstraintColumns(subject.RelatedColumns, expectation.RelatedColumns);
+            }
+        }
+
+        private static void CompareConstraintColumns(DataColumn[] subjectColumns, DataColumn[] expectationColumns)
+        {
+            var subjectColumnNames = new HashSet<string>(subjectColumns.Select(col => col.ColumnName));
+            var expectationColumnNames = new HashSet<string>(expectationColumns.Select(col => col.ColumnName));
+
+            var allColumnNames = new HashSet<string>();
+
+            allColumnNames.UnionWith(subjectColumnNames);
+            allColumnNames.UnionWith(expectationColumnNames);
+
+            foreach (var columnName in allColumnNames)
```

Comment:
If I read this right, it's making tautological checks, i.e. if `subjectColumnNames` contains all column names from `subjectColumnNames`.

Would it work to do this:

```
columnsOnlyInSubject = subjectColumns.Except(expectationColumns);
columnsOnlyInExpected = expectationColumns.Except(subjectColumns);

using new AssertionScope();

foreach ( var column in columnsOnlyInSubject)
{
    // FailWith
}

foreach ( var column in columnsOnlyInExpected)
{
    // FailWith
}
```

Reply:
Yep, I think it would. What's the `using new AssertionScope()` line about?

Reply:
I have done some refactoring, generalizing the `JoinUsingWritingStyle` method from `TimeSpanValueFormatter` into an extension method in the process. `JoinUsingWritingStyle` is rewritten to avoid enumerating the parameter multiple times. The failure message now lists all columns contributing to failure rather than just the first one it finds.